### PR TITLE
feat(notifications): native UI updates

### PR DESCRIPTION
**Pull request by Calinteodor**
Updates notifications ui and flow on mobile.

### CODE REVIEW

File: react/features/base/icons/svg/index.ts [hunk 1]
```diff
@@ -47,7 +47,6 @@ export { default as IconHelp } from './help.svg';
 export { default as IconHighlight } from './highlight.svg';
 export { default as IconImage } from './image.svg';
 export { default as IconInfoCircle } from './info-circle.svg';
-export { default as IconBurger } from './burger.svg';
```
Comment (saghul) [inline react/features/base/icons/svg/index.ts:50]: Are we using the icon at all? If not, can we delete it?

File: react/features/base/ui/components/native/Button.tsx [hunk 2]
```diff
@@ -28,13 +29,17 @@ const Button: React.FC<IProps> = ({
     labelKey,
     labelStyle,
     onClick: onPress,
+    rippleColor = true,
```
Comment (saghul) [inline react/features/base/ui/components/native/Button.tsx:32]: This prop name suggests a color, not a boolean. How about `useRippleColor` ?

File: react/features/lobby/middleware.ts [hunk 3]
```diff
@@ -13,6 +15,11 @@ import StateListenerRegistry from '../base/redux/StateListenerRegistry';
 import { playSound, registerSound, unregisterSound } from '../base/sounds/actions';
 import { isTestModeEnabled } from '../base/testing/functions';
 import { handleLobbyChatInitialized, removeLobbyChatParticipant } from '../chat/actions.any';
+import { navigate }
```
Comment (saghul) [inline react/features/lobby/middleware.ts:18]: This is going to break web. We can't import platform dependent code in a platform agnostic middleware...

File: react/features/lobby/middleware.ts [hunk 4]
```diff
@@ -245,24 +208,45 @@ function _handleLobbyNotification(store: IStore) {
             dispatch(hideNotification(LOBBY_NOTIFICATION_ID));
             dispatch(rejectKnockingParticipant(firstParticipant.id));
         }) ];
-    } else {
+
+        if (showChat) {
+            customActionNameKey.splice(1, 0, 'lobby.chat');
```
Comment (saghul) [inline react/features/lobby/middleware.ts:213]: This reads really weird, I'd say it needs an explanatory comment.

File: react/features/lobby/middleware.ts [hunk 5]
```diff
@@ -245,24 +208,45 @@ function _handleLobbyNotification(store: IStore) {
...
             waitingParticipants: knockingParticipants.length
         });
         icon = NOTIFICATION_ICON.PARTICIPANTS;
+        concatText = true;
         customActionNameKey = [ 'notify.viewLobby' ];
         customActionHandler = [ () => batch(() => {
             dispatch(hideNotification(LOBBY_NOTIFICATION_ID));
+
+            if (navigator.product === 'ReactNative') {
+                navigate(screen.conference.participants);
```
Comment (saghul) [inline react/features/lobby/middleware.ts:236]: We should probably provide a native implementation for `openParticipantsPane` which does the apprpriate navigation, instead on needed to have platform dependent code here.

File: react/features/notifications/components/native/Notification.tsx [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+const ICON_COLOR = {
+    error: colors.error06,
+    normal: colors.primary06,
+    success: colors.success05,
+    warning: colors.warning05
+};
+
+
+type Props = AbstractNotificationProps & WithTranslation & {
+    _participants: ArrayLike<any>;
```
Comment (saghul) [inline react/features/notifications/components/native/Notification.tsx:60]: This is really weird. Why does a notification need to have a reference to an array of participants?

File: react/features/notifications/components/native/Notification.tsx [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+            customActionNameKey
+            // @ts-ignore
+        } = this.props;
+
+        if (customActionNameKey?.length && customActionHandler?.length) {
+            return customActionNameKey?.map((customAction: string, index: number) => {
+                let buttonLabelStyle: Object | undefined;
+
+                if (customAction === 'lobby.reject') {
+                    buttonLabelStyle = styles.rejectBtnLabel;
```
Comment (saghul) [inline react/features/notifications/components/native/Notification.tsx:90]: Why doesn't the web need this level of trickery? A generic notification component should not know about what its content is. Everything needs to come via props.

File: react/features/notifications/components/native/Notification.tsx [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+
+    /**
+     * Implements React's {@link Component#render()}.
+     *
+     * @inheritdoc
+     * @returns {ReactElement}
+     */
+    render() {
+        // @ts-ignore
+        const { _participants, icon } = this.props;
```
Comment (saghul) [inline react/features/notifications/components/native/Notification.tsx:181]: This really needs to go away...

File: react/features/notifications/components/native/Notification.tsx [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+        const { _participants } = this.props;
+
+        return (
+            <>
+                {
+                    _participants?.map((p: any, index: number) => (
+                        <View
+                            key = { index }
+                            style = { styles.avatarContainer }>
+                            <Avatar
```
Comment (saghul) [inline react/features/notifications/components/native/Notification.tsx:264]: Why are we displaying avatars here? The web just renders the participant icon, since we don't have a real avatar anyway, and they'd be too small to be meaningful. Please drop this.

File: react/features/notifications/components/native/Notification.tsx [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+                                { p.name }
+                            </Text>
+                        </View>
+                    ))
+                }
+            </>
+        );
+    }
+
+    _getDescription: () => Array<string>;
```
Comment (saghul) [inline react/features/notifications/components/native/Notification.tsx:280]: I don't think this is necessary now that this is TS.

File: react/features/notifications/reducer.ts [hunk 11]
```diff
@@ -90,6 +90,12 @@ ReducerRegistry.register<INotificationsState>('features/notifications',
  * queue.
  */
 function _insertNotificationByPriority(notifications: INotification[], notification: INotification) {
+
+    // We need to filter out the notifications from queue that have the same id as the new notifications,
```
Comment (saghul) [inline react/features/notifications/reducer.ts:94]: I think before filtering we should get the index, because if the new notification has a given UUID we want it to replace the existing notification, in the exact same position.

File: react/features/notifications/reducer.ts [hunk 12]
```diff
@@ -112,7 +118,7 @@ function _insertNotificationByPriority(notifications: INotification[], notificat
     }
 
     // Create a copy to avoid mutation and insert the notification.
-    const copyOfNotifications = notifications.slice();
+    const copyOfNotifications = filteredNotifications.slice();
```
Comment (saghul) [inline react/features/notifications/reducer.ts:121]: This is no longer necessary since the `filter()` operation will have created a new array.

File: react/features/participants-pane/components/native/ContextMenuLobbyParticipantReject.js [hunk 13]
```diff
@@ -35,7 +35,6 @@ const ContextMenuLobbyParticipantReject = ({ participant: p }: Props) => {
         <View
             style = { styles.contextMenuItemSectionAvatar }>
             <Avatar
-                className = 'participant-avatar'
```
Comment (saghul) [inline react/features/participants-pane/components/native/ContextMenuLobbyParticipantReject.js:38]: Why is this change necessary?

**Review by saghul — CHANGES_REQUESTED** [review]
Left some comments, PTAL! I noticed there are no changes to `NotificationsContainer`, but we most certainly need changes there, since it currently only renders one notification at a time. Will follow-up with some UI feedback.

**Comment by saghul** [comment]
UI / UX feedback:

- Please adjust NotificationContainer to render all notifications, like web does. We might even be able to unify them into a single component, please check it out.
- I did not see the "X participants joined the meeting" notification get updated, did you?
- Can we add some fade-in/out effects to the notifications, like web has?
- I saw a glitch in rendering the icon:

IMG_6982

File: react/features/base/icons/svg/index.ts (hunk 1 above)
Comment (Calinteodor) [reply react/features/base/icons/svg/index.ts:50]: The icon is not being used, that's why I deleted it.

File: react/features/base/ui/components/native/Button.tsx (hunk 2 above)
Comment (Calinteodor) [reply react/features/base/ui/components/native/Button.tsx:32]: I agree, will change.

File: react/features/lobby/middleware.ts (hunk 4 above)
Comment (Calinteodor) [reply react/features/lobby/middleware.ts:213]: Like, this checks if lobby chat button is available and if so, it adds it to the customActionNameKey array?

File: react/features/lobby/middleware.ts (hunk 5 above)
Comment (Calinteodor) [reply react/features/lobby/middleware.ts:236]: Ok, I will create an action that does the navigation on native and not depend on platform type.

File: react/features/notifications/components/native/Notification.tsx (hunk 6 above)
Comment (Calinteodor) [reply react/features/notifications/components/native/Notification.tsx:60]: I use this for the avatar to show. We show it only for one participant lobby notification.

File: react/features/notifications/components/native/Notification.tsx (hunk 7 above)
Comment (Calinteodor) [reply react/features/notifications/components/native/Notification.tsx:90]: Because web uses atlaskit.

File: react/features/notifications/components/native/Notification.tsx (hunk 9 above)
Comment (Calinteodor) [reply react/features/notifications/components/native/Notification.tsx:264]: Latest design shows avatar inside notifications.

File: react/features/notifications/components/native/Notification.tsx (hunk 7 above)
Comment (saghul) [reply react/features/notifications/components/native/Notification.tsx:90]: Ah bummer. Can you come up with a way to pass this as props please?

File: react/features/notifications/components/native/Notification.tsx (hunk 9 above)
Comment (saghul) [reply react/features/notifications/components/native/Notification.tsx:264]: Let's not. It's so small it's useless.

File: react/features/notifications/components/native/Notification.tsx (hunk 10 above)
Comment (Calinteodor) [reply react/features/notifications/components/native/Notification.tsx:280]: Will check if it is so.

File: react/features/notifications/reducer.ts (hunk 11 above)
Comment (Calinteodor) [reply react/features/notifications/reducer.ts:94]: Will test this.

File: react/features/notifications/components/native/Notification.tsx (hunk 6 above)
Comment (saghul) [reply react/features/notifications/components/native/Notification.tsx:60]: Mentioned below, let's drop the avatar please.

File: react/features/notifications/reducer.ts (hunk 12 above)
Comment (Calinteodor) [reply react/features/notifications/reducer.ts:121]: Correct. Will update.

File: react/features/lobby/middleware.ts (hunk 4 above)
Comment (saghul) [reply react/features/lobby/middleware.ts:213]: Yeah, an explanation of _what_ we are doing with that splice.

File: react/features/participants-pane/components/native/ContextMenuLobbyParticipantReject.js (hunk 13 above)
Comment (Calinteodor) [reply react/features/participants-pane/components/native/ContextMenuLobbyParticipantReject.js:38]: Because we don't use classNames on native. This was added in a separate commit.

File: react/features/base/icons/svg/index.ts (hunk 1 above)
Comment (saghul) [reply react/features/base/icons/svg/index.ts:50]: Cool! By delete i meant the SVG too.

Comment (Calinteodor) [reply react/features/base/icons/svg/index.ts:50]: Oh, ok. You are talking about the actual file. I will delete it, yes.

File: react/features/lobby/middleware.ts (hunk 5 above)
Comment (Calinteodor) [reply react/features/lobby/middleware.ts:236]: And I think I will have to do the same for the action dispatched when lobby chat button is pressed.

File: react/features/notifications/components/native/Notification.tsx (hunk 7 above)
Comment (Calinteodor) [reply react/features/notifications/components/native/Notification.tsx:90]: Will tackle this deeper, yes.

File: react/features/participants-pane/components/native/ContextMenuLobbyParticipantReject.js (hunk 13 above)
Comment (saghul) [reply react/features/participants-pane/components/native/ContextMenuLobbyParticipantReject.js:38]: 👍

File: react/features/lobby/middleware.ts (hunk 5 above)
Comment (saghul) [reply react/features/lobby/middleware.ts:236]: 👍

File: react/features/lobby/middleware.ts [hunk 14]
```diff
@@ -245,24 +202,40 @@ function _handleLobbyNotification(store: IStore) {
...
+        }
+    }
+
+    if (knockingParticipants.length > 1) {
         descriptionKey = 'notify.participantsWantToJoin';
         notificationTitle = i18n.t('notify.waitingParticipants', {
             waitingParticipants: knockingParticipants.length
         });
         icon = NOTIFICATION_ICON.PARTICIPANTS;
+        concatText = true;
```
Comment (saghul) [inline react/features/lobby/middleware.ts:224]: We should drop `concatText` now, since it was a workaround for the mobile notifications we used to have, since they didn't have the context of title and description.

File: react/features/notifications/components/native/Notification.tsx [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+// @ts-ignore
+import styles from './styles';
+
+
+/**
+ * Default value for the maxLines prop.
+ *
+ * @type {number}
+ */
+const DEFAULT_MAX_LINES = 2;
```
Comment (saghul) [inline react/features/notifications/components/native/Notification.tsx:43]: Why do we want to be so restrictive?

File: react/features/participants-pane/actions.native.ts [hunk 16]
```diff
@@ -103,3 +93,15 @@ export function showRoomParticipantMenu(room: Object, participantJid: string, pa
...
 }
+
+/**
+ * Action to open the participants pane.
+ *
+ * @returns {Object}
+ */
+export const open = () => {
+    return {
+        participantsPaneRoute: navigate(screen.conference.participants),
```
Comment (saghul) [inline react/features/participants-pane/actions.native.ts:104]: You are actually navigating here, not storing the route. Is that intentional?

File: react/features/chat/actions.native.ts [hunk 17]
```diff
@@ -6,15 +14,30 @@ export * from './actions.any';
...
+ *     participant: participant,
+ *     privateChatRoute,
  *     type: OPEN_CHAT
  * }}
  */
-export function openChat(participant: Object) {
+export function openChat(participant: Object, pollsDisabled: boolean) {
     return {
+        lobbyChatRoute: pollsDisabled
+            ? navigate(screen.conference.chat)
```
Comment (saghul) [inline react/features/chat/actions.native.ts:29]: You are actually performing a navigation here, is that intentional?

File: react/features/base/ui/components/native/Button.tsx [hunk 18]
```diff
@@ -27,31 +29,36 @@ const Button: React.FC<IProps> = ({
...
-    type
+    type,
+    useRippleColor = true
 }: IProps) => {
     const { t } = useTranslation();
-    const { CONTAINED } = BUTTON_MODES;
     const { DESTRUCTIVE, PRIMARY, SECONDARY, TERTIARY } = BUTTON_TYPES;
+    const { CONTAINED, TEXT } = BUTTON_MODES;
+
+    const enableRippleColor
```
Comment (saghul) [inline react/features/base/ui/components/native/Button.tsx:42]: Can you rename this to just `rippleColor`? I find it easier to make sense of.

File: react/features/base/ui/constants.any.ts [hunk 19]
```diff
@@ -13,6 +13,8 @@ export enum BUTTON_TYPES {
  */
 export const BUTTON_MODES: {
     CONTAINED: 'contained';
+    TEXT: 'text';
 } = {
-    CONTAINED: 'contained'
+    CONTAINED: 'contained',
```
Comment (saghul) [inline react/features/base/ui/constants.any.ts:18]: ISss there no better way to annotate this?

File: react/features/chat/actions.web.ts [hunk 20]
```diff
@@ -13,7 +12,7 @@ export * from './actions.any';
  *
  * @param {Object} participant - The recipient for the private chat.
  * @returns {{
- *     participant: Participant,
+ *     participant: participant,
```
Comment (saghul) [inline react/features/chat/actions.web.ts:15]: Why this change?

File: react/features/notifications/components/native/NotificationsContainer.js [hunk 21]
```diff
@@ -72,6 +71,15 @@ class NotificationsContainer extends Component<Props> {
     componentDidMount() {
         // Set the initial dismiss timeout (if any)
         this._manageDismissTimeout();
+
+        Animated.timing(
```
Comment (saghul) [inline react/features/notifications/components/native/NotificationsContainer.js:75]: Let's put the animations in the individual notifications, so they can all fade in.

File: react/features/notifications/reducer.ts [hunk 22]
```diff
@@ -90,30 +90,41 @@ ReducerRegistry.register<INotificationsState>('features/notifications',
  * queue.
  */
 function _insertNotificationByPriority(notifications: INotification[], notification: INotification) {
+
+    // Create a copy to avoid mutation.
+    const copyOfNotifications = notifications.slice();
+
+    // Check if any queued notification has the same id as the new notification
+    const sameUIDNotification = (queuedNotification: INotification) => queuedNotification?.uid === notification?.uid;
```
Comment (saghul) [inline react/features/notifications/reducer.ts:98]: Doesn't it look cleaner to inline this in the findIndex call? No big deal though.

File: react/features/notifications/reducer.ts [hunk 23]
```diff
@@ -90,30 +90,41 @@ ReducerRegistry.register<INotificationsState>('features/notifications',
...
+
+    // Get the index
+    let insertAtLocation = copyOfNotifications.findIndex(sameUIDNotification);
+
+    if (insertAtLocation !== -1) {
+
```
Comment (saghul) [inline react/features/notifications/reducer.ts:104]: style: unnecessary newline

**Review by saghul — CHANGES_REQUESTED** [review]
Left some comments, I'll take it for a spin now.

File: react/features/chat/actions.web.ts (hunk 20 above)
Comment (Calinteodor) [reply react/features/chat/actions.web.ts:15]: Reverted.

File: react/features/base/ui/constants.any.ts (hunk 19 above)
Comment (Calinteodor) [reply react/features/base/ui/constants.any.ts:18]: Can you be more specific, please? I don't get it.

Comment (saghul) [reply react/features/base/ui/constants.any.ts:18]: The type description is bacially the same as the content. I wonder if there is no other way around it?

Comment (Calinteodor) [reply react/features/base/ui/constants.any.ts:18]: We use react-native-paper for buttons and these are the modes that Button component accepts.
https://callstack.github.io/react-native-paper/button.html

**Review by saghul — APPROVED** [review]
